# Project description

<div class="alert alert-info"> <b>

    You've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition             for a new product. 

# Libraries

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = 'all'

In [2]:
#database
db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# Studying the tables

In [3]:
# reading tables
books = pd.read_sql('SELECT * FROM books', con=engine)
authors = pd.read_sql('SELECT * FROM authors', con=engine)
publishers = pd.read_sql('SELECT * FROM publishers', con=engine)
ratings = pd.read_sql('SELECT * FROM ratings', con=engine)
reviews = pd.read_sql('SELECT * FROM reviews', con=engine)

In [4]:
# books table
books.sample(3)
print()
books.info()

book_id  author_id                                              title  \
737      738        473                The Good Earth (House of Earth  #1)   
93        94        205  Batman: Arkham Asylum - A Serious House on Ser...   
615      616        213                                        Tell No One   

     num_pages publication_date  publisher_id  
737        418       2009-03-04           158  
93         216       2005-01-11            71  
615        370       2009-08-25            82


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [5]:
# authors table
authors.sample(3)
print()
authors.info()

author_id                                  author
379        380  Maeve Binchy/René Huigen/Frans Thomése
175        176      Frances Hodgson Burnett/Nancy Bond
211        212             H.G. Wells/Arthur C. Clarke


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [6]:
# publishers table
publishers.sample(3)
print()
publishers.info()

publisher_id             publisher
157           158  Howard Publishing Co
138           139             HarperOne
210           211    Penguin (Business)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [7]:
# ratings table
ratings.sample(3)
print()
ratings.info()

rating_id  book_id      username  rating
3443       3444      549  abigailmoore       4
1702       1703      299       dharris       4
6250       6251      959      robert00       5


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [8]:
# reviews table
reviews.sample(3)
print()
reviews.info()

review_id  book_id       username  \
2591       2591      925   jasongoodman   
2410       2410      864   webbmichelle   
1716       1715      623  perrypatricia   

                                                   text  
2591  Work improve late team letter perform color. R...  
2410  White discuss skin open rule reality role. Spr...  
1716                               Eye perform pattern.


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


<div class="alert alert-info"> <b>

    after showing samples of 5 dataset given, looks like all the data is complete contains no missing value and sorted      well. the datasets contains many details about books were released with book_id, dates, text, reviews and more.

**Find the number of books released after January 1, 2000.**

In [9]:
# sample of the data
books.sample(2)

book_id  author_id                title  num_pages publication_date  \
735      736        231  The Gods Themselves        288       2000-02-10   
863      864        137   The Secret History        559       2004-04-13   

     publisher_id  
735           182  
863           309

In [10]:
# query
query = '''
        SELECT COUNT(*)
        FROM books
        WHERE publication_date > '2000-01-01'
'''
filtered_books = pd.io.sql.read_sql(query, con = engine)

In [11]:
# result
filtered_books

count
0    819

In [12]:
print('books released after January 1, 2000:', filtered_books.iloc[0, 0])

books released after January 1, 2000: 819


<div class="alert alert-info"> <b>

    the number of books released after January 1, 2000 is 819 books.
    the data contains 1000 rows and 6 columns with the books details

**Find the number of user reviews and the average rating for each book.**

In [13]:
ratings.sample(2)
print()
reviews.sample(2)

rating_id  book_id  username  rating
1795       1796      300  samuel96       3
1223       1224      207   grogers       3

review_id  book_id        username  \
1198       1198      437      martinadam   
1101       1102      406  williamswesley   

                                                   text  
1198  Billion care suffer. Thank data improve southe...  
1101  Sea only anything social figure music. Machine...

In [14]:
# query
query1 = ''' 
    SELECT
        books.title AS book_title,
        COUNT(DISTINCT reviews.review_id) AS num_reviews,
        AVG(ratings.rating) AS avg_rating
    FROM books
LEFT JOIN reviews ON reviews.book_id = books.book_id
LEFT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY books.title
ORDER BY avg_rating DESC;
'''
avg_rate_review = pd.io.sql.read_sql(query1, con = engine)

In [15]:
# result for all books with number of reviews and avg rating
avg_rate_review = avg_rate_review.sort_values(by='num_reviews', ascending = False)
avg_rate_review.head(10)
avg_rate_review

book_title  num_reviews  \
328                                Memoirs of a Geisha            8   
688                            Twilight (Twilight  #1)            7   
599             The Da Vinci Code (Robert Langdon  #2)            6   
256  Harry Potter and the Chamber of Secrets (Harry...            6   
600                             The Catcher in the Rye            6   
564                                Water for Elephants            6   
614                                      The Alchemist            6   
295                                   The Glass Castle            6   
836                                    Eat  Pray  Love            6   
620                                           The Road            6   

     avg_rating  
328    4.138462  
688    3.662500  
599    3.830508  
256    4.287500  
600    3.825581  
564    3.977273  
614    3.789474  
295    4.206897  
836    3.395833  
620    3.772727

book_title  num_reviews  \
328                                Memoirs of a Geisha            8   
688                            Twilight (Twilight  #1)            7   
599             The Da Vinci Code (Robert Langdon  #2)            6   
256  Harry Potter and the Chamber of Secrets (Harry...            6   
600                             The Catcher in the Rye            6   
..                                                 ...          ...   
479                               Leonardo's Notebooks            0   
388                          Essential Tales and Poems            0   
907                            The Natural Way to Draw            0   
652    Anne Rice's The Vampire Lestat: A Graphic Novel            0   
3     The Cat in the Hat and Other Dr. Seuss Favorites            0   

     avg_rating  
328    4.138462  
688    3.662500  
599    3.830508  
256    4.287500  
600    3.825581  
..          ...  
479    4.000000  
388    4.000000  
907    3.000000  
652    3.666667  
3      5.000000  

[999 rows x 3 columns]

<div class="alert alert-info"> <b>

    Here we can see the Top ten books by the number of reviews, while the scale of rating is 1 to 5 it looks like these books are really good because somehow it made some people leave a review while other books that had 5 in rating                   score had the same time 0 review.

**Identify the publisher that has released the greatest number of books with more than 50 pages**

In [16]:
books.sample(2)
publishers.sample(2)

book_id  author_id        title  num_pages publication_date  publisher_id
424      425        383  Meditations        303       2006-04-27           212
848      849        445   The Rescue        352       2005-04-01           116

publisher_id publisher
284           285  Starfire
196           197      Onyx

In [17]:
# query
query2 = '''
    SELECT 
        publishers.publisher,
        COUNT(*) as num_books
    FROM books
JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE num_pages > 50
GROUP BY publishers.publisher
ORDER BY num_books DESC;
'''
top_publishers = pd.io.sql.read_sql(query2, con = engine)

In [18]:
# result
top_publishers.head(3)
print("publisher with the highest number of books with more than 50 pages is:", top_publishers['publisher'].iloc[0])

publisher  num_books
0             Penguin Books         42
1                   Vintage         31
2  Grand Central Publishing         25

publisher with the highest number of books with more than 50 pages is: Penguin Books


<div class="alert alert-info"> <b>

    here we can see top publishers who released the highest number of books with more than 50 pages.
    the Top publisher is is Penguin Books

**Identify the author with the highest average book rating**

In [19]:
ratings.sample()

rating_id  book_id  username  rating
4132       4133      673  dwayne09       4

In [20]:
# query
query3 = '''
    SELECT
        authors.author,
        AVG(ratings.rating) as avg_rating
    FROM authors
JOIN books ON authors.author_id = books.author_id
JOIN ratings ON books.book_id = ratings.book_id
WHERE books.book_id IN 
    (SELECT book_id 
     FROM ratings 
     GROUP BY book_id 
     HAVING COUNT(*) >= 50)
GROUP BY authors.author
ORDER BY avg_rating DESC;
'''
highest_author = pd.io.sql.read_sql(query3, con = engine)

In [21]:
# result
highest_author.head(3)
print("author with the highest average book rating is:", highest_author['author'].iloc[0])

author  avg_rating
0         J.K. Rowling/Mary GrandPré    4.287097
1  Markus Zusak/Cao Xuân Việt Khương    4.264151
2                     J.R.R. Tolkien    4.246914

author with the highest average book rating is: J.K. Rowling/Mary GrandPré


<div class="alert alert-info"> <b>

    here we can see above the top three authors with the highest average book rating with 4.2 rate of 5

**Find the average number of text reviews among users who rated more than 50 books**

In [22]:
books.sample()

book_id  author_id                            title  num_pages  \
198      199        435  Dream Country (The Sandman  #3)        111   

    publication_date  publisher_id  
198       1999-02-05            71

In [23]:
query4 = '''
    SELECT 
    AVG(reviews_count) AS avg_reviews_count
    FROM
    (SELECT
    COUNT(*) AS reviews_count
    FROM ratings r
    JOIN reviews rev ON r.book_id = rev.book_id AND r.username = rev.username
    WHERE r.username IN(
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(*) > 50)
GROUP BY r.username
) AS subquery;
'''
text_reviews = pd.io.sql.read_sql(query4, con = engine)

In [24]:
text_reviews
text_reviews = round(text_reviews['avg_reviews_count'][0], 2)
print("average number of text reviews among users who rated more than 50 books is:", text_reviews)

avg_reviews_count
0          24.333333

average number of text reviews among users who rated more than 50 books is: 24.33


<div class="alert alert-info"> <b>

    the number of text reviews amongs users is 24.33 for those users who have more than 50 books.